### ISSUE TITLE
Installer Failure on MacOS 10.14

### ISSUE THREAD

**Issue by HackerJL** [NagiosEnterprises/ncpa#491]
Upgraded a machine to new MacOS 10.14.

NCPA was not running so I installed it again.  Result:

$ sudo zsh install.sh 
Starting upgrade...

Stopping old NCPA services...done

Re-loading services...done

Setting permissions...chown: nagios: illegal user name

**Comment by jomann09** [comment]
Did you change the user that ncpa runs as afterwards? I don't think on upgrade it creates the nagios user/group although it tries to set the new ncpa files to the proper user after install.

There is actually going to be a slight change to this in the future too, so the files won't be owned by nagios, but instead by root.

Also, to finish the install, run:
```
cp -Rf /tmp/ncpa_etc /usr/local/ncpa

launchctl load /Library/LaunchDaemons/com.nagios.ncpa.listener.plist
launchctl load /Library/LaunchDaemons/com.nagios.ncpa.passive.plist

launchctl start com.nagios.ncpa.passive
launchctl start com.nagios.ncpa.listener
```

**Comment by HackerJL** [comment]
Thanks, ill give those commands a try later today and report back.

To answer your question, its a one-user Mac mini (username default) with the base NCPA installed, no extra fancy stuff (so no, didn't change the user).  After Mojave install, I could see it wasn't working and issued the command as I pasted.

Ill let you know how it goes.

**Comment by jomann09** [comment]
Oh I see so you installed it again, there's an interesting issue I've seen in the past where it won't create the user, so if you check your users and don't see a "nagios" user you might need to create one for it to run properly.

**Comment by HackerJL** [comment]
Alright, for the sake of simplicity, is there a command I should use to just uninstall so it does it fresh and proper on its own install?

**Comment by jomann09** [comment]
It doesn't look like you made it to the section that copies over the uninstall, you should be able to uninstall by running `sh /usr/local/ncpa/uninstall.sh` there should also be an uninstall script in the mounted `NCPA-x.x.x` directory.

**Comment by HackerJL** [comment]
great, ill check it out and report back.

**Comment by HackerJL** [comment]
Uninstall worked, thanks for that.  

Install seems to stop short.  
$ sudo zsh install.sh

Starting upgrade...

Stopping old NCPA services...$

Tried the uninstall from the mounted folder.  doesn't seem to really finish anything.  Oddly.

**Comment by jomann09** [comment]
The one from the mounted folder I would expect wouldn't work from there, you'd want to move it into `/usr/local/ncpa` before running it. Is the ncpa folder still there? The install says it's upgrading which is odd.

**Comment by HackerJL** [comment]
Alright, so Im still having some stress about it, but let me lay out what ive tried.

- There is nothing in the /usr/local/ncpa folder except the Library (and inside it the preference folder - which is empty).  And I cannot delete it. As root or otherwise, operation not permitted.

- Copying the uninstaller there does nothing different I'm afraid.  

- Reboot after just in case, no different.  Installer still says starting upgrade, and ends the same way.  No new files copied or written to /usr/local/ncpa

-After the reboot, I tried to remove the Library folder, no change.  So I did the install thinking its as clean as it gets.  This is the output (I also suck at markdown apparently):

sudo zsh install.sh 
Starting upgrade...
Stopping old NCPA services...done
Re-loading services...done
Setting permissions...cp: /usr/local/ncpa/etc: No such file or directory

**Comment by jomann09** [comment]
It's possible that the user creation part of the script doesn't work, if you reset perms, make sure the directory is empty and create a nagios user/group manually and re-run the install will it work?

In order to have it use root you just have to edit your ncpa.cfg file. You can do that in /usr/local/ncpa/etc/ncpa.cfg you should see uid and gid settings.

You could also do the above to set the user to root in the config after you install it by commenting out the parts in the install that set the user permissions.

**Comment by HackerJL** [comment]
I stripped out all the create user and group in the installer, and the installer finished.  Hurrah.  

I went and edited the NCPA.cfg file and set uid to root, gid to wheel.  

No matter what, I cant seem to get it to load properly.  System.log shows me the following:
com.apple.xpc.launchd[1] (com.nagios.ncpa.listener[40446]): Service exited with abnormal code: 1

I cant seem to find anymore nerdy wording to help explain anything to that point.  

Anything else you'd like me to try?

**Comment by jomann09** [comment]
What happens when you run `/usr/local/ncpa/ncpa_listener -n` as root from the cmdline?

**Comment by HackerJL** [comment]
It runs!...but with errors below.  But its working, the API is live.  

sh-3.2# ./ncpa_listener -n
2018-10-02 14:47:44,858 56409 ERROR too many groups
Traceback (most recent call last):
  File "/Users/ncpa/ncpa/agent/ncpadaemon.py", line 261, in set_uid_gid
ValueError: too many groups
2018-10-02 14:47:44,868 56409 INFO started
2018-10-02 14:47:44,868 56409 INFO Using SSL version TLSv1
2018-10-02 14:48:21,938 56409 INFO Did not receive normal values. Unable to find meaningful check.

**Comment by HackerJL** [comment]
So...this tells me something is maybe up with the launcher. 

Wondering if Im not launching as root, but instead as my user.  Without sudo I guess.  Let me see.

**Comment by HackerJL** [comment]
Yup, it was just me not launching it properly as root.  Reboot helped as they were launched by root, not my user.  So, to report, this is working good now.  As long as its installed by root, it's good.  

So, the installer, I stripped out lines 33 to 71.

After that, changed the uid,gid in the ncpa.cfg file to root/wheel and away it went.

Many thanks, I hope I helped test what needs/needed to be fixed in next version.
